scheduler/scheduler-server/src/test/java/functionaltests/recover/RunningTaskRecoveryWithDownNodeTest.java

```diff
@@ -0,0 +1,219 @@
+    @Test
+    public void action() throws Throwable {
+
+        this.createNodes();
+
+        JobId jobid = this.submitJob();
+
+        this.waitForAllTasksToRun(jobid);
+
+        this.killSchedulerAndNodes();
+
+        long timeForSchedulerToBeUp = this.recordSchedulerRestartTime();
+
+        Scheduler scheduler = schedulerHelper.getSchedulerInterface();
+        JobState jobState = scheduler.getJobState(jobid);
+
+        this.checkTasks(jobState);
+
+        this.waitForJobToFinish(jobid);
+
+        this.checkJobResult(jobid, scheduler);
+
+        this.checkSchedulerStateRecoveryDoesNotWaitTaskPingAttemptTimesFrequency(timeForSchedulerToBeUp);
+    }
+
+    private long recordSchedulerRestartTime() throws Exception {
+        long schedulerStartTime = System.currentTimeMillis();
+        this.restartScheduler();
+        long schedulerUpTime = System.currentTimeMillis();
+        return schedulerUpTime - schedulerStartTime;
+    }
+
+    private void checkSchedulerStateRecoveryDoesNotWaitTaskPingAttemptTimesFrequency(long timeForSchedulerToBeUp) {
+        int attempts = PASchedulerProperties.SCHEDULER_NODE_PING_ATTEMPTS.getValueAsInt();
+        int frequency = PASchedulerProperties.SCHEDULER_NODE_PING_FREQUENCY.getValueAsInt();
+        int minimumTotalRecoveryDurationIfRetryMechanismIsExecuted = attempts * frequency * 1000 * NUMBER_OF_TASKS;
+
+        assertThat((int) timeForSchedulerToBeUp).isLessThan(minimumTotalRecoveryDurationIfRetryMechanismIsExecuted);
+    }
+
+    private void checkJobResult(JobId jobid, Scheduler scheduler)
+            throws NotConnectedException, PermissionException, UnknownJobException {
+        JobResult jobResult = scheduler.getJobResult(jobid);
+        Assert.assertFalse(jobResult.hadException());
+    }
+
+    private void waitForJobToFinish(JobId jobid) throws Exception {
+        schedulerHelper.waitForEventJobFinished(jobid);
+    }
+
+    private void checkTasks(JobState jobState) {
+        int maximumNumberOfPendingTasks = jobState.getNumberOfPendingTasks();
+
+        for (int i = 0; i < maximumNumberOfPendingTasks; i++) {
+
+            TaskState taskState = jobState.getTasks().get(i);
+            TaskStatus taskStatus = taskState.getTaskInfo().getStatus();
+
+            assertThat(taskStatus.equals(TaskStatus.PENDING) || taskStatus.equals(TaskStatus.RUNNING)).isTrue();
+        }
+    }
+
+    private void restartScheduler() throws Exception {
+        schedulerHelper = new SchedulerTHelper(false,
+                                               new File(SCHEDULER_CONFIGURATION_RESTART.toURI()).getAbsolutePath(),
+                                               new File(RM_CONFIGURATION_RESTART.toURI()).getAbsolutePath(),
+                                               null,
+                                               false);
+        this.createNodes();
+    }
+
+    private void killSchedulerAndNodes() throws Exception {
+        Thread.sleep(RESTART_SCHEDULER_INTER_TIME_MILLIS);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this use of "Thread.sleep()". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2925)